### PR TITLE
Gutenframe: Simplify redirect logic

### PR DESCRIPTION
**Pull request by noahtallen**
### Changes proposed in this Pull Request
This is motivated by my recent editor experience. After #51370 was merged, I noticed I was being redirected to wp-admin 90% of the time on simple sites. This does resolve that problem, and I tried to simplify this logic as well -- we were redirecting in different cases and had some logic in one place which should have been in a different place.

#### When do we currently redirect:
When looking into this, I noticed that we are doing a redirect in two places. First, we set up a timeout in `componentDidMount` which redirects after 25 seconds. This redirect always happens unless the timeout is cleared. Currently, that timeout is cleared using the `onLoad` prop for the actual `iframe` element. _Previously_ (a week or two ago) we actually had _another_ timeout in the `onLoad` method which waited for several seconds until the `WindowActions.Loaded` message was received by the `onMessage` handler in this component, and only then did it clear the timeout. That timeout was originally added to fix an IE11 bug where it always redirected to wp-admin. That was removed in #51370 because we no longer support IE11.

In other words, we currently redirect in two cases:
1. If the iframe hasn't loaded at all after 25 seconds
2. If the iframe has "loaded" but the [iframe-bridge-server didn't send the "loaded" message](https://github.com/Automattic/wp-calypso/blob/194f92fdbbd5275ac360601c121d1710ae6785e8/apps/wpcom-block-editor/src/calypso/features/iframe-bridge-server.js#L1127). (This happens instantly once iframe.onLoad is called, it doesn't wait around for the bridge server.)

#### Resulting bug:
The cause of the constant redirects is because the `onLoad` iframe callback is being executed _before_ the iframe bridge server initializes. This was previously not an issue because we had a timeout which waited for the bridge server to load. Though we aren't subject to network issues here, it's just not guaranteed that the iframe bridge server will execute before iframe.onload is called. iframe.onload just means that the server has loaded all the scripts and content in the frame.

#### Complexity:
My intuition here is to remove a lot of this complexity. If we are waiting to do anything until the iframe bridge server sends us a message, we should put more of our code _there_ rather than in the `iframe.onLoad` handler. I think this will give us more accurate performance measurements and state updates as well.

So currently, I've created a single timer to track when the editor has loaded. There are two stages of the timer:
1. The first stage waits 25 seconds from when we start loading the iframe.
2. The second stage clears the first stage and then waits 1 more second for the iframe-bridge-server to respond.

If the timer runs out in either case, we redirect to wp-admin assuming a failure to load the iframe.
  
### Testing instructions
1. Open calypso.live
2. Visit the iframed editor of a simple site in firefox.
3. It should not redirect. (Try this on several refreshes too.)

### CODE REVIEW

File: client/gutenberg/editor/calypsoify-iframe.tsx
Comment (ciampo) [inline client/gutenberg/editor/calypsoify-iframe.tsx]: If `this.props.shouldLoadIframe === false` we potentially never clear the interval?

Maybe we should consider moving the `if ( this.props.shouldLoadIframe ) {` check outside the `setInterval` call?

File: client/gutenberg/editor/calypsoify-iframe.tsx
Comment (ciampo) [inline client/gutenberg/editor/calypsoify-iframe.tsx]: Nit: the full stop mark was probably intended to be after `timer` in the next line?

File: client/gutenberg/editor/calypsoify-iframe.tsx
Comment (simison) [inline client/gutenberg/editor/calypsoify-iframe.tsx]: Any chance to optimize loading speed here by waiting only for DOM to load, and not CSS/other assets? See notes at https://developer.mozilla.org/en-US/docs/Web/API/GlobalEventHandlers/onload

File: client/gutenberg/editor/calypsoify-iframe.tsx
Comment (simison) [inline client/gutenberg/editor/calypsoify-iframe.tsx]: Would `onerror` and `onabort` event handlers catch these faster than a timer?

Comment (noahtallen) [reply client/gutenberg/editor/calypsoify-iframe.tsx]: Good thought! I was looking for something like that but didn't find anything. Will investigate

File: client/gutenberg/editor/calypsoify-iframe.tsx
Comment (noahtallen) [reply client/gutenberg/editor/calypsoify-iframe.tsx]: The reason this exists is to detect if the iframe loads a page, but it isn't the editor (e.g. 404 or unauthenticated page). In that case, it's not easy to detect those errors inside the iframe. So we're relying on the iframe bridge server post message to tell us when the editor page exists. In my testing, that takes a few hundred ms or so.

This wouldn't save any time on the happy path because all of the logic dealing with showing the editor happens after receiving the message from the iframe bridge server. And that's kind of how it has to be because the calypso integration won't work unless the bridge server is initialized.

At any rate, I don't _think_ we have been introducing any artificial delay. The redirect timer is always cleared immediately after we detect that the editor has loaded

File: client/gutenberg/editor/calypsoify-iframe.tsx
Comment (noahtallen) [reply client/gutenberg/editor/calypsoify-iframe.tsx]: I'm not sure what the original intent for this code was, but I think the existing logic &: to basically run the interval _until_ `shouldLoadIframe` is true. In other words, we don't want to start a countdown if shouldLoadIframe is false, but we do as soon as it's true.

If this were a functional component, it'd be a good candidate for useEffect. I'm not sure if componentDidUpdate would replicate it entirely

Comment (simison) [reply client/gutenberg/editor/calypsoify-iframe.tsx]: Just FYI, the [`sandbox`](https://github.com/WordPress/gutenberg/blob/3da717b8d0ac7d7821fc6d0475695ccf3ae2829f/packages/components/src/sandbox/index.js) component in the core might be a good inspiration for component structure.

File: client/gutenberg/editor/calypsoify-iframe.tsx
Comment (simison) [reply client/gutenberg/editor/calypsoify-iframe.tsx]: Gotcha, thanks for elaborating!

File: client/gutenberg/editor/calypsoify-iframe.tsx
Comment (noahtallen) [reply client/gutenberg/editor/calypsoify-iframe.tsx]: @simison I'm no iframe expert, but my best understanding of the situation is that those don't really work for iframes, particularly cross-origin iframes.

File: client/gutenberg/editor/calypsoify-iframe.tsx
Comment (noahtallen) [reply client/gutenberg/editor/calypsoify-iframe.tsx]: thanks!

File: client/gutenberg/editor/calypsoify-iframe.tsx
Comment (noahtallen) [reply client/gutenberg/editor/calypsoify-iframe.tsx]: I've refactored this into didMount/didUpdate. Hopefully we can refactor this to a FC in the future

File: client/gutenberg/editor/calypsoify-iframe.tsx
Comment (noahtallen) [inline client/gutenberg/editor/calypsoify-iframe.tsx:172]: @glendaviesnz @scruffian @aaronrobertshaw Could you perhaps comment on these changes? Do we still need to use this lengthy redirect approach or have the underlying auth issues been solved?

The main change is instead of having multiple timers, I've implemented one timer:
1. up to 25 seconds until iframe.onload is called
2. a further 1 second until the iframe bridge server sends us a response

What do you think?

File: client/gutenberg/editor/calypsoify-iframe.tsx
Comment (ciampo) [inline client/gutenberg/editor/calypsoify-iframe.tsx]: Is the `2.5s` on purpose, or was it meant to be `25s`?

File: client/gutenberg/editor/calypsoify-iframe.tsx
Comment (glendaviesnz) [reply client/gutenberg/editor/calypsoify-iframe.tsx:172]: Certainly wouldn't hurt to simplify this code if possible. From memory some of these checks were added to handle the loading of Atomic sites with custom domains, and [jetpack sites with iframe headers set](https://github.com/Automattic/wp-calypso/pull/36977), so any changes would need testing against these contexts rather than just testing that simple sites load the iframed editor.

File: client/gutenberg/editor/calypsoify-iframe.tsx
Comment (noahtallen) [reply client/gutenberg/editor/calypsoify-iframe.tsx]: thanks for catching this!

File: client/gutenberg/editor/calypsoify-iframe.tsx
Comment (noahtallen) [reply client/gutenberg/editor/calypsoify-iframe.tsx:172]: Thanks for the context! I just tested with a JN site, and it looks like the X-Frame redirect would happen after `iframe.onload` is called. Additionally, the auth redirect happens at a higher layer of the stack before trying to load the iframe component. (Ran into that when authenticating and setting up wpcom SSO)

Screen Shot 2021-04-13 at 2 17 41 PM

I think this makes sense to me because `iframe.onload` is called when the frame's sources have finished loading. So if the browser shows an error page like above, the frame still considers it as loading a page. Which is kind of ridiculous, but that seems to be pretty consistent behavior, and matches articles I'm reading about cross-origin iFrames.

I'll do some testing with custom Atomic domains too, but if this holds up, we can probably remove the 25 second redirect timer and just utilize `iframe.onload`

Comment (noahtallen) [reply client/gutenberg/editor/calypsoify-iframe.tsx:172]: Another benefit of that is `iframe.onload` fires only when all scripts and content have been downloaded by the frame, so we wouldn't need an extra long timeout to handle network connections

Comment (noahtallen) [reply client/gutenberg/editor/calypsoify-iframe.tsx:172]: Though, looking at #43248, it might be important to keep for the auth redirect problem 🤔 

Though if that problem is solved, we won't need it either

Comment (noahtallen) [reply client/gutenberg/editor/calypsoify-iframe.tsx:172]: Yep, I'll just keep the long redirect around to handle cases like that. I'll try to be clear in the comments that that's a more rare condition

**Comment by noahtallen** [comment]
Alright, this is ready to go. IMO, it should be easier to maintain down the road because everything is more explicit. Additionally, the comments should make it easier for future maintainers to understand the inner workings of the redirect timers.

**Comment by yansern** [comment]
Iframe redirection due to slow network connections has been a pain point for me. Thanks for getting this sorted!

If I were to add one more thing, it would be to have a config flag that lets devs disable redirection altogether. Not a blocker for this PR.

**Review by yansern — APPROVED** [review]

**Comment by noahtallen** [comment]
I'm noting that the e2e failure is not because tests are failing, but because of the "metric change" failure. It thinks that more tests should be passing, but those tests have been removed or muted on trunk since this PR was created. So it should be safe to bypass.

File: client/gutenberg/editor/calypsoify-iframe.tsx
Comment (ciampo) [inline client/gutenberg/editor/calypsoify-iframe.tsx:147]: When looking at code changes file in my IDE, I get a TypeScript error on this line:

`'"calypso/lib/performance-tracking"' has no exported member named 'PerformanceTrackProps'.`

Although I don't think that this error was introduced by this PR, it would be still good to remove this and the remaining TypeScript errors — probably best done in a follow-up PR (including the `/* eslint-disable @typescript-eslint/no-explicit-any */` at the top of the file)

Comment (noahtallen) [reply client/gutenberg/editor/calypsoify-iframe.tsx:147]: good idea 👍

**Comment by noahtallen** [comment]
This didn't end up working: I was being redirected every time on staging after merging. Will have to investigate that more

**Pull request merged**
